osutil/kcmdline.go

```diff
@@ -204,6 +204,116 @@ func KernelCommandLineKeyValues(keys ...string) (map[string]string, error) {
 	return m, nil
 }
 
+// KernelValidateParam validates a kernel parameter and splits it in
+// key and value in case there is no error.
+// See lib/cmdline.c:next_arg in kernel sources.
+
+// KernelArgument represents a parsed kernel argument
+type KernelArgument struct {
+	Param string
+	Value string
+}
+
+// KernelParseCommandline parses a kernel command line, returning a
+// slice with the arguments in the same order as in cmdline. Note that
+// kernel arguments can be repeated. We follow the same algorithm as in
+// linux kernel's function lib/cmdline.c:next_arg as far as possible.
+// TODO Replace KernelCommandLineSplit with this eventually
+func KernelParseCommandline(cmdline string) (args []KernelArgument) {
+	cmdlineBy := []byte(cmdline)
+	args = []KernelArgument{}
+	start := firstNotSpace(cmdlineBy)
+	for start < len(cmdlineBy) {
+		param, val, end := parseArgument(cmdlineBy[start:])
+		// Hyphens (dashes) and underscores are equivalent in parameter names,
+		// so we standardize in "_".
+		param = strings.ReplaceAll(param, "-", "_")
+		args = append(args, KernelArgument{Param: param, Value: val})
+		start += end
+		start += firstNotSpace(cmdlineBy[start:])
+	}
+
+	return args
+}
+
+// Does the same as isspace() in tools/include/nolibc/ctype.h from the
+// linux kernel
+func isSpace(b byte) bool {
+	switch b {
+	case ' ', '\t', '\n', '\v', '\f', '\r':
+		return true
+	}
+	return false
+}
+
+// Similar to skip_spaces() in lib/string_helpers.c from the linux kernel
+func firstNotSpace(args []byte) int {
+	var i int
+	var b byte
+	for i, b = range args {
+		if !isSpace(b) {
+			return i
+		}
+	}
+	return i + 1
+}
+
+// parseArgument parses a kernel argument that is known to start at
+// the beginning of args, returning the parameter and the assigned
+// value if any, plus where the argument ends in args.
+//
+// This follows the same algorithm as the next_arg function from
+// lib/cmdline.c in the linux kernel, to make sure we handle the
+// arguments in exactly the same way.
+func parseArgument(args []byte) (param, val string, end int) {
+	var i, equals, startArg int
+	var argQuoted, inQuote bool
+
+	if args[0] == '"' {
+		startArg++
+		argQuoted = true
+		inQuote = true
+	}
+
+	for i = startArg; i < len(args); i++ {
+		if isSpace(args[i]) && !inQuote {
+			break
+		}
+		if args[i] == '=' && equals == 0 {
+			equals = i
+		}
+		if args[i] == '"' {
+			inQuote = !inQuote
+		}
+	}
+
+	end = i
+	endParam := i
+	var subsVal, subsParam int
```

Comment:
what does subs stand for here?

Reply:
I've added a comment trying to clarify. Note again that the behavior is peculiar and I'm trying to replicate exactly what the kernel does - in the end the kernel implementation is very limited and parameters are expected to follow some implicit rules, otherwise you probably will get some random strings if there are unmatched quotes, etc.